### PR TITLE
Add `blocked_main_thread` to sync file spans

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Add `blocked_main_thread` to sync file spans

## :bulb: Motivation and Context

Relates to #1727

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by denrase** [comment]
@buenaflor I'm not sure we can guarantee that the `debugName` will always return `main` on the main isolate. AFAIK there is no other way to check in dart only code.

**Comment by buenaflor** [comment]
@denrase do we know any scenarios where it won't work?

**Comment by ueman** [comment]
> @denrase do we know any scenarios where it won't work?

Probably when someone calls [`PlatformDispatcher.setIsolateDebugName()`](https://api.flutter.dev/flutter/dart-ui/PlatformDispatcher/setIsolateDebugName.html) in the main isolate from Flutter, but it's probably not that likely. I'm not sure for Dart only code though.

**Comment by buenaflor** [comment]
Got it, If we can't guarantee that it's always correct maybe it should be opt-in?

> How to proceed with backend implementation?

https://docs.sentry.io/product/issues/issue-details/performance-issues/file-main-thread-io/ 

it should already be in place and all we need is to add the correct span data. If I looked correctly it should be 

`blocked_main_thread: true/false`

examples: [java](https://github.com/getsentry/sentry-java/blob/606074f713f0cb84a5d3e6dc8d0adef592ce7281/sentry/src/main/java/io/sentry/instrumentation/file/FileIOSpanManager.java#L106), [cocoa](https://github.com/getsentry/sentry-cocoa/blob/42ef6baf82435c9e60f92974097657417ca4839a/Sources/Sentry/SentryNSDataTracker.m#L193C15-L193C15)

File: file/lib/src/sentry_file.dart
Comment (buenaflor) [inline file/lib/src/sentry_file.dart]: this span data field should be different I think, check my other comment

**Review by buenaflor — CHANGES_REQUESTED** [review]

**Comment by denrase** [comment]
@buenaflor Think i found a good solution. Added a callback to options to check for the main Isolate. When using the `dart` SDK we use the `debugName`, and on flutter we use `ServicesBinding.rootIsolateToken` to determine if we are on the main Isolate. So for flutter this should always be correct, and in dart only it should be correct as long as 'main' is returned as a value.

**Comment by buenaflor** [comment]
`ServicesBinding.rootIsolateToken` is only available on Flutter >= 3.7.0. 
See [here](https://docs.flutter.dev/release/release-notes/release-notes-3.7.0): look for https://github.com/flutter/flutter/pull/111712

I guess we need to fall back to `debugName` if someone is using a lower Flutter version. 

Maybe we can leverage the dynamic - NoSuchMethod approach again here?

**Comment by denrase** [comment]
@buenaflor It looks like this won't work in this case, as this is a static property on on `ServicesBinding`. So we don't have an instance which we can cast to `dynamic`.

Is there another way to check this? 🤔

If not, it looks like we can only fall back to the Isolate debug name.

**Comment by denrase** [comment]
@buenaflor Removed `ServicesBinding.rootIsolateToken` check for now, since I'm not aware how to call static API that is not present in older flutter versions. Maybe @marandaneto has an idea?

**Comment by marandaneto** [comment]
> @buenaflor Removed `ServicesBinding.rootIsolateToken` check for now, since I'm not aware how to call static API that is not present in older flutter versions. Maybe @marandaneto has an idea?

All `rootIsolateToken` does is to call `RootIsolateToken.instance` and this class and property [exists in older versions](https://github.com/flutter/engine/pull/35174/files), right?
All you need is to `result = RootIsolateToken.instance != null` I guess?

**Comment by denrase** [comment]
@marandaneto  I checked and`RootIsolateToken.instance` was introduced to `flutter/engine` on 8. September 2022 as far as I can tell. I did not find any direct reference to it in release notes in the repos.

https://github.com/flutter/engine/pull/35174

Flutter `3.3.2`, released on `14.9.2022`, was the first flutter release after this PR merge.

So I'm guessig that flutter versions < `3.3.2` do not have this API, and per our oubspec.yaml we support Flutter >= 3.0.0.

I have not seen the class definition in other repositories, expept in `dart-lang/flute`, which is used for benchmarking.

I wanted to test this and downgrade to Flutter 3, but with no luck so far on macOS.

**Comment by marandaneto** [comment]
> @marandaneto I checked and`RootIsolateToken.instance` was introduced to `flutter/engine` on 8. September 2022 as far as I can tell. I did not find any direct reference to it in release notes in the repos.
> 
> [flutter/engine#35174](https://github.com/flutter/engine/pull/35174)
> 
> Flutter `3.3.2`, released on `14.9.2022`, was the first flutter release after this PR merge.
> 
> So I'm guessig that flutter versions < `3.3.2` do not have this API, and per our oubspec.yaml we support Flutter >= 3.0.0.
> 
> I have not seen the class definition in other repositories, expept in `dart-lang/flute`, which is used for benchmarking.
> 
> I wanted to test this and downgrade to Flutter 3, but with no luck so far on macOS.

https://github.com/leoafarias/fvm

**Comment by denrase** [comment]
Like expected, we do not have access to `RootIsolateToken`. Installed Flutter version `3.3.10`. Had to remove several sentry plugins from the example app, as those do not all support the same low SDK versions as the main package to finally get the compilation error.

```
denis@Deniss-MBP example % fvm doctor

FVM Version: 2.4.1
___________________________________________________

FVM config found:
___________________________________________________

Project: example
Directory: /Users/denis/Repos/sentry/sentry-dart/flutter/example
Version: 3.3.10
Project Flavor: None selected
___________________________________________________

Version is currently cached locally.

Cache Path: /Users/denis/fvm/versions/3.3.10
Channel: false
SDK Version: 3.3.10

IDE Links
VSCode: .fvm/flutter_sdk
Android Studio: /Users/denis/Repos/sentry/sentry-dart/flutter/example/.fvm/flutter_sdk


Configured env paths:
___________________________________________________

Flutter:
/Users/denis/Repos/flutter/bin/flutter

Dart:
/opt/homebrew/bin/dart

FVM_HOME:
not set

denis@Deniss-MBP example % fvm flutter run -d 6B9B6A68-2804-4447-9959-3CCAB49E9501
Launching lib/main.dart on iPhone 15 Pro in debug mode...
Running Xcode build...                                          
Xcode build done.                                            8,5s
Failed to build iOS app
Error output from Xcode build:
↳
    --- xcodebuild: WARNING: Using the first of multiple matching destinations:
    { platform:iOS Simulator, id:6B9B6A68-2804-4447-9959-3CCAB49E9501, OS:17.2, name:iPhone 15 Pro }
    { platform:iOS Simulator, id:6B9B6A68-2804-4447-9959-3CCAB49E9501, OS:17.2, name:iPhone 15 Pro }
    ** BUILD FAILED **


Xcode's output:
↳
    Writing result bundle at path:
    	/var/folders/_n/s4t83kqj137f_9465c5z63k40000gn/T/flutter_tools.3jOctp/flutter_ios_build_temp_dirGarRkE/temporary_xcresult_bundle

    ../lib/src/sentry_flutter.dart:127:14: Error: Undefined name 'RootIsolateToken'.
          return RootIsolateToken.instance != null;
                 ^^^^^^^^^^^^^^^^
    Failed to package /Users/denis/Repos/sentry/sentry-dart/flutter/example.
    Command PhaseScriptExecution failed with a nonzero exit code
    note: Run script build phase 'Run Script' will be run during every build because the option to run the script phase "Based on dependency analysis" is unchecked. (in target 'Runner' from project 'Runner')
    note: Run script build phase 'Thin Binary' will be run during every build because the option to run the script phase "Based on dependency analysis" is unchecked. (in target 'Runner' from project 'Runner')

Could not build the application for the simulator.
Error launching application on iPhone 15 Pro.
denis@Deniss-MBP example % 
```

**Comment by marandaneto** [comment]
> RootIsolateToken.instance

@denrase background isolates is Flutter >= [3.7](https://medium.com/flutter/whats-new-in-flutter-3-7-38cbea71133c)
So in this case, either do not use this class/field and do not set the `blocked_main_thread` at all or bump the min. Flutter version.
`Isolate.current.debugName` is mutable so it can lead to false positives since it can be changed.

File: file/test/sentry_file_test.dart
Comment (marandaneto) [inline file/test/sentry_file_test.dart:661]: All tests are using the `isMainIsolate` fake but not starting a real background isolate and reading a file from it to verify that indeed returns `true` or `false` correctly.
So it is hard to tell that the solution works.
I'd write a test that creates a background isolate and wait for its result, and asserting that the span contains the right result.
This test could be done on the flutter integration test for example if needs to be, or even the example app, since there the flutter min. version is higher or can be higher

**Comment by denrase** [comment]
@marandaneto Exactly.

- Can't use `[rootIsolateToken](https://api.flutter.dev/flutter/services/ServicesBinding/rootIsolateToken.html)` without bumping the Flutter SDK.
- It is a static and we can't use the `(instance as dynamic).anyMethod` approach for versions < `3.7`
- Using `Isolate.current.debugName` is not reliable.

Only way as I see it to reliably and automatically detect this is to bump to Flutter `3.7`

File: file/test/sentry_file_test.dart
Comment (denrase) [reply file/test/sentry_file_test.dart:661]: Added two tests to the example app. The first runs on the main isolate and the second init's sentry only on a bg isolate and runs there. Results as expected, thx for the suggestion! 👍

**Comment by denrase** [comment]
@buenaflor The PR still has the caveat that users might override the `debugName` we are using.

https://github.com/getsentry/sentry-dart/issues/1727#issuecomment-1840756267

**Comment by buenaflor** [comment]
Let's keep this blocked by flutter min 3.7 wdyt? @denrase

**Comment by denrase** [comment]
@buenaflor Sounds good to me. 👍

**Pull request closed without merging**
